### PR TITLE
[GPU] (I)Dft with single layer test

### PR DESCRIPTION
**Pull request by serhii-pavlovskyi-altran**
on top of some unrelated fixes

naming convention check is failing due to issue in unrelated code, which was hidden by bug fixed in first commit

### CODE REVIEW

File: src/tests/functional/plugin/gpu/shared_tests_instances/single_layer_tests/dft.cpp [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+//
+
+#include <single_layer_tests/dft.hpp>
+#include <common_test_utils/test_constants.hpp>
+
+namespace {
+
+const std::initializer_list<ngraph::helpers::DFTOpType> opTypes {
+    ngraph::helpers::DFTOpType::FORWARD,
+//    ngraph::helpers::DFTOpType::INVERSE
```
Comment (tgubanova-lohika) [inline src/tests/functional/plugin/gpu/shared_tests_instances/single_layer_tests/dft.cpp:12]: is it expected to be commented?

File: src/tests/functional/plugin/gpu/shared_tests_instances/single_layer_tests/range.cpp [hunk 2]
```diff
@@ -21,15 +21,15 @@ const std::initializer_list<InferenceEngine::Precision> netPrecisions {
 
 using LayerTestsDefinitions::RangeLayerTest;
 INSTANTIATE_TEST_SUITE_P(smoke_Basic, RangeLayerTest,
```
Comment (tgubanova-lohika) [inline src/tests/functional/plugin/gpu/shared_tests_instances/single_layer_tests/range.cpp:23]: should it be part of this PR?

File: src/tests/functional/shared_test_classes/src/single_layer/dft.cpp [hunk 3]
```diff
@@ -5,6 +5,18 @@
 #include "shared_test_classes/single_layer/dft.hpp"
 
 namespace LayerTestsDefinitions {
+static bool invalidTestSuffixChar(char c) {
+    switch (c) {
+    case '.':
+//    case '-':
+//    case '=':
+//    case '(':
+//    case ')':
```
Comment (tgubanova-lohika) [inline src/tests/functional/shared_test_classes/src/single_layer/dft.cpp:14]: should it be removed?

File: src/plugins/intel_gpu/src/plugin/ops/dft.cpp [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+        auto r = outShape.size();
+        if (r != 5) // only tested on 5 and bfzyx is hardcoded below
+            throw std::runtime_error { "dft v7 output rank is " + std::to_string(r) };
+    }
+    auto axes = dynamic_cast<const ngraph::op::Constant&>(*op->get_input_node_shared_ptr(1)).cast_vector<int32_t>();
+    {
+        auto dataRank = op->get_input_shape(0).size();
+        for (auto &axis : axes)
+            if (axis < 0)
+                axis += dataRank - 1;
```
Comment (tgubanova-lohika) [inline src/plugins/intel_gpu/src/plugin/ops/dft.cpp:29]: I think sth like "normalizeAxes" already presented in common utils. In any case better to move it to a separate method for readability

File: cmake/developer_package/ncc_naming_style/ncc_naming_style.cmake [hunk 5]
```diff
@@ -107,8 +107,10 @@ function(ov_ncc_naming_style)
 
     list(APPEND NCC_STYLE_ADDITIONAL_INCLUDE_DIRECTORIES "${NCC_STYLE_SOURCE_DIRECTORY}")
 
+# without it sources with same name from different directories will map to same .ncc_style target
+    file(RELATIVE_PATH source_dir_rel ${CMAKE_SOURCE_DIR} ${NCC_STYLE_SOURCE_DIRECTORY})
     foreach(source IN LISTS sources)
-        set(output_file "${ncc_style_bin_dir}/${source}.ncc_style")
+        set(output_file "${ncc_style_bin_dir}/${source_dir_rel}/${source}.ncc_style")
```
Comment (tgubanova-lohika) [inline cmake/developer_package/ncc_naming_style/ncc_naming_style.cmake:113]: could you investigate if naming conv bot failed because of your changes? In any case better to move it to the separate PR

File: src/plugins/intel_gpu/src/plugin/ops/dft.cpp (hunk 4 above)
Comment (tgubanova-lohika) [reply src/plugins/intel_gpu/src/plugin/ops/dft.cpp:29]: sth like this I mean
https://github.com/openvinotoolkit/openvino/blob/3bd6eb6dcdeeea1b1381354cff3a9c814a1a77ee/src/core/src/validation_util.cpp#L884

File: src/plugins/intel_gpu/src/kernel_selector/core/cl_kernels/dft_ref.cl [hunk 6]
```diff
@@ -0,0 +1,163 @@
+// Copyright (C) 2021 Intel Corporation
+// SPDX-License-Identifier: Apache-2.0
+//
+
+typedef float2 cfloat;
+#if 0
```
Comment (tgubanova-lohika) [inline src/plugins/intel_gpu/src/kernel_selector/core/cl_kernels/dft_ref.cl:6]: I propose either to add comments on why we are using macroses instead of methods with correspond benchmarks or just leave inline methods.

File: src/tests/functional/plugin/gpu/shared_tests_instances/single_layer_tests/dft.cpp (hunk 1 above)
Comment (serhii-pavlovskyi-altran) [reply src/tests/functional/plugin/gpu/shared_tests_instances/single_layer_tests/dft.cpp:12]: it will be implemented as part of next pr for idft

File: src/tests/functional/plugin/gpu/shared_tests_instances/single_layer_tests/range.cpp (hunk 2 above)
Comment (serhii-pavlovskyi-altran) [reply src/tests/functional/plugin/gpu/shared_tests_instances/single_layer_tests/range.cpp:23]: i can split it if needed

File: src/tests/functional/shared_test_classes/src/single_layer/dft.cpp (hunk 3 above)
Comment (serhii-pavlovskyi-altran) [reply src/tests/functional/shared_test_classes/src/single_layer/dft.cpp:14]: technically those all are incorrect symbols, but they don't break eclipse parser. maybe we need some global decision

File: cmake/developer_package/ncc_naming_style/ncc_naming_style.cmake (hunk 5 above)
Comment (serhii-pavlovskyi-altran) [reply cmake/developer_package/ncc_naming_style/ncc_naming_style.cmake:113]: my change uncovered ncc check failure. i can move it to separate pr, but i can't build openvino locally without this fix

File: src/tests/functional/plugin/gpu/shared_tests_instances/single_layer_tests/dft.cpp (hunk 1 above)
Comment (tgubanova-lohika) [reply src/tests/functional/plugin/gpu/shared_tests_instances/single_layer_tests/dft.cpp:12]: could you either remove it and add in next PR or add comments about this with TODO - as now it looks little bit confusing

File: src/tests/functional/plugin/gpu/shared_tests_instances/single_layer_tests/range.cpp (hunk 2 above)
Comment (tgubanova-lohika) [reply src/tests/functional/plugin/gpu/shared_tests_instances/single_layer_tests/range.cpp:23]: if it's not about DFT at all - better to split

File: src/plugins/intel_gpu/src/kernel_selector/core/cl_kernels/dft_ref.cl [hunk 7]
```diff
@@ -0,0 +1,163 @@
...
+}
+inline cfloat czero() {
+    return (cfloat)(0);
+}
+#else
+#define real(a) ((a).s0)
+#define imag(a) ((a).s1)
+#define cmult(a, b) ((cfloat)(real(a) * real(b) - imag(a) * imag(b), real(a) * imag(b) + imag(a) * real(b)))
+#define cadd(a, b) ((cfloat)(real(a) + real(b), imag(a) + imag(b)))
+#define expi(x) ((cfloat)(cos(x), sin(x)))
```
Comment (tgubanova-lohika) [inline src/plugins/intel_gpu/src/kernel_selector/core/cl_kernels/dft_ref.cl:40]: as I remember it's expected to have undefs for all the defines in the end of cl file

Comment (serhii-pavlovskyi-altran) [reply src/plugins/intel_gpu/src/kernel_selector/core/cl_kernels/dft_ref.cl:40]: these are generc complex primitives, they can be put in common header and used by any kernel working with complex numbers

Comment (tgubanova-lohika) [reply src/plugins/intel_gpu/src/kernel_selector/core/cl_kernels/dft_ref.cl:40]: so we either need to put them to the common header or leave them only here but with expected rules

File: src/plugins/intel_gpu/src/kernel_selector/core/cl_kernels/dft_ref.cl (hunk 6 above)
Comment (serhii-pavlovskyi-altran) [reply src/plugins/intel_gpu/src/kernel_selector/core/cl_kernels/dft_ref.cl:6]: i'd like to leave one implementation after gathering feedback

File: src/tests/functional/plugin/gpu/shared_tests_instances/single_layer_tests/range.cpp (hunk 2 above)
Comment (serhii-pavlovskyi-altran) [reply src/tests/functional/plugin/gpu/shared_tests_instances/single_layer_tests/range.cpp:23]: while possible, it will not be easy, because both range and dft depend on https://github.com/openvinotoolkit/openvino/pull/9891/commits/ceb0e55747febcabb7aca74722117a9eb092c350

File: src/tests/functional/plugin/gpu/shared_tests_instances/single_layer_tests/dft.cpp (hunk 1 above)
Comment (serhii-pavlovskyi-altran) [reply src/tests/functional/plugin/gpu/shared_tests_instances/single_layer_tests/dft.cpp:12]: added comment

File: src/plugins/intel_gpu/src/plugin/ops/dft.cpp (hunk 4 above)
Comment (serhii-pavlovskyi-altran) [reply src/plugins/intel_gpu/src/plugin/ops/dft.cpp:29]: used different overload

File: src/plugins/intel_gpu/src/kernel_selector/core/cl_kernels/dft_ref.cl (hunk 7 above)
Comment (serhii-pavlovskyi-altran) [reply src/plugins/intel_gpu/src/kernel_selector/core/cl_kernels/dft_ref.cl:40]: i'll investigate how it's supposed to work

File: src/plugins/intel_gpu/src/kernel_selector/core/cl_kernels/dft_ref.cl (hunk 6 above)
Comment (serhii-pavlovskyi-altran) [reply src/plugins/intel_gpu/src/kernel_selector/core/cl_kernels/dft_ref.cl]: functions require some macro help to support fp16 in addition to fp32, i've dropped functions

File: src/tests/functional/plugin/gpu/shared_tests_instances/single_layer_tests/dft.cpp (hunk 1 above)
Comment (serhii-pavlovskyi-altran) [reply src/tests/functional/plugin/gpu/shared_tests_instances/single_layer_tests/dft.cpp]: implemented idft

File: src/plugins/intel_gpu/src/kernel_selector/core/cl_kernels/dft_ref.cl (hunk 7 above)
Comment (serhii-pavlovskyi-altran) [reply src/plugins/intel_gpu/src/kernel_selector/core/cl_kernels/dft_ref.cl:12]: undefined macros

File: src/plugins/intel_gpu/include/intel_gpu/primitives/dft.hpp [hunk 8]
```diff
@@ -0,0 +1,29 @@
+// Copyright (C) 2021 Intel Corporation
```
Comment (yeonbok) [inline src/plugins/intel_gpu/include/intel_gpu/primitives/dft.hpp:1]: 2021-2022

File: src/plugins/intel_gpu/src/kernel_selector/core/cl_kernels/dft_ref.cl [hunk 9]
```diff
@@ -0,0 +1,140 @@
...
+{
+    size_t g2 = get_global_id(2);
+    ptrdiff_t y1 = g2 % OUTPUT_SIZES[1];
+#if OUTPUT_DIMS > 2
+    g2 /= OUTPUT_SIZES[1];
+    ptrdiff_t y2 = g2 % OUTPUT_SIZES[2];
+#if OUTPUT_DIMS > 3
+    g2 /= OUTPUT_SIZES[2];
+    ptrdiff_t y3 = g2 % OUTPUT_SIZES[3];
+#if OUTPUT_DIMS > 4
```
Comment (yeonbok) [inline src/plugins/intel_gpu/src/kernel_selector/core/cl_kernels/dft_ref.cl:29]: its only supporting bfzyx, do we need to divide these cases?

File: src/plugins/intel_gpu/include/intel_gpu/primitives/dft.hpp (hunk 8 above)
Comment (serhii-pavlovskyi-altran) [reply src/plugins/intel_gpu/include/intel_gpu/primitives/dft.hpp]: done

File: src/plugins/intel_gpu/src/kernel_selector/core/cl_kernels/dft_ref.cl (hunk 9 above)
Comment (serhii-pavlovskyi-altran) [reply src/plugins/intel_gpu/src/kernel_selector/core/cl_kernels/dft_ref.cl:29]: tests only test one rank, cl kernel supports fewer dimensions and can be easily extended to support more. if it's known for sure that we will never need other ranks, i can get rid of some ifdefs in kernel

**Review by yeonbok — APPROVED** [review]

File: src/core/include/openvino/core/validation_util.hpp [hunk 10]
```diff
@@ -113,7 +113,7 @@ int64_t normalize_axis(const std::string& node_description,
 /// \param[in]  axes              The requested vector of axes.
 ///
 OPENVINO_API
-void normalize_axes(const Node* node, const int64_t& tensor_rank, std::vector<int64_t>& axes);
+void normalize_axes(const Node* node, int64_t tensor_rank, std::vector<int64_t>& axes);
```
Comment (yeonbok) [inline src/core/include/openvino/core/validation_util.hpp:116]: can I ask why this is changed to non const type? seems there is no reason. If there is no critical reason, could you please revert the unnecessary change?

File: src/plugins/intel_gpu/src/plugin/ops/dft.cpp [hunk 11]
```diff
@@ -0,0 +1,49 @@
...
+    auto axes = dynamic_cast<const ngraph::op::Constant&>(*op->get_input_node_shared_ptr(1)).cast_vector<int64_t>();
+    {
+        auto dataRank = op->get_input_shape(0).size();
+        ov::normalize_axes(op.get(), dataRank - 1, axes);
+    }
+    auto outTensor = tensor_from_dims(outShape);
+    cldnn::data_types outDataType = DataTypeFromPrecision(op->get_output_element_type(0));
+    cldnn::layout outLayout { outDataType, cldnn::format::bfzyx, outTensor };
+    cldnn::dft prim { layer_type_name_ID(op), move(p.GetInputPrimitiveIDs(op)[0]), move(axes), outLayout,
+        kind, op->get_friendly_name() };
```
Comment (yeonbok) [inline src/plugins/intel_gpu/src/plugin/ops/dft.cpp:33]: why the outLayout should be decided here? Please remove output layout setup from here and set it from the calc_output_layout as other primitive. And then remove output layout from pritmivie.

Comment (yeonbok) [reply src/plugins/intel_gpu/src/plugin/ops/dft.cpp:33]: Also it shouldnt be always bfzyx, should it? please calculate according to the inputs.

File: src/tests/functional/shared_test_classes/src/base/layer_test_utils.cpp [hunk 12]
```diff
@@ -464,8 +464,8 @@ void LayerTestsCommon::Validate() {
     if (functionRefs == nullptr) {
         functionRefs = ngraph::clone_function(*function);
     }
-    auto expectedOutputs = CalculateRefs();
-    const auto &actualOutputs = GetOutputs();
+    const auto expectedOutputs = CalculateRefs();
+    const auto actualOutputs = GetOutputs();
```
Comment (yeonbok) [inline src/tests/functional/shared_test_classes/src/base/layer_test_utils.cpp:468]: Please do not apply unnecessary change for the common code.

File: src/tests/functional/shared_test_classes/src/single_layer/dft.cpp [hunk 13]
```diff
@@ -22,26 +34,26 @@ std::string DFTLayerTest::getTestCaseName(const testing::TestParamInfo<DFTParams
     result << "SignalSize=" << CommonTestUtils::vec2str(signalSize) << "_";
     result << "Inverse=" << (opType == ngraph::helpers::DFTOpType::INVERSE) << "_";
     result << "TargetDevice=" << targetDevice;
-    return result.str();
+    auto str = result.str();
+    replace_if(str.begin(), str.end(), invalidTestSuffixChar, '_');
```
Comment (yeonbok) [inline src/tests/functional/shared_test_classes/src/single_layer/dft.cpp:38]: Is this necessary?

File: src/tests/functional/shared_test_classes/src/single_layer/dft.cpp [hunk 14]
```diff
@@ -22,26 +34,26 @@ std::string DFTLayerTest::getTestCaseName(const testing::TestParamInfo<DFTParams
...
-
+    ngraph::ParameterVector params { std::make_shared<ngraph::opset1::Parameter>(inType, ov::Shape { inputShape }) };
+    auto& data = params.front();
+    data->set_friendly_name("data");
+    auto dft = ngraph::builder::makeDFT(data, axes, signalSize, opType);
 
     ngraph::ResultVector results{std::make_shared<ngraph::opset1::Result>(dft)};
-    function = std::make_shared<ngraph::Function>(results, paramVector, "DFT");
+    function = std::make_shared<ngraph::Function>(results, params, "DFT");
+    threshold = 0.05;
```
Comment (yeonbok) [inline src/tests/functional/shared_test_classes/src/single_layer/dft.cpp:57]: where is this used?

File: src/plugins/intel_gpu/include/intel_gpu/primitives/dft.hpp [hunk 15]
```diff
@@ -0,0 +1,29 @@
...
+
+#pragma once
+
+#include "primitive.hpp"
+
+namespace cldnn {
+struct dft: public primitive_base<dft> {
+    CLDNN_DECLARE_PRIMITIVE(dft)
+
+    enum Kind {
```
Comment (yeonbok) [inline src/plugins/intel_gpu/include/intel_gpu/primitives/dft.hpp:13]: Could you rename this more specific to this operation like DftKind?

File: src/plugins/intel_gpu/include/intel_gpu/primitives/dft.hpp [hunk 16]
```diff
@@ -0,0 +1,29 @@
...
+
+    dft(const primitive_id &id, primitive_id &&data, std::vector<int64_t> &&axes, const layout &output_layout,
+        Kind kind, const primitive_id &ext_prim_id = { }) :
+            primitive_base { id, { move(data) }, ext_prim_id, output_layout.data_padding },
+            axes { move(axes) },
+            output_layout { output_layout },
+            kind { kind } {
+    }
+    std::vector<int64_t> axes;
+    layout output_layout;
```
Comment (yeonbok) [inline src/plugins/intel_gpu/include/intel_gpu/primitives/dft.hpp:26]: Please get rid of output_layout from primitive and let it to be set from program_node

File: src/plugins/intel_gpu/src/kernel_selector/core/actual_kernels/dft/dft_kernel_ref.cpp [hunk 17]
```diff
@@ -0,0 +1,91 @@
...
+}
+
+ParamsKey DFTKernelRef::GetSupportedKey() const {
+    ParamsKey k;
+    k.EnableInputDataType(Datatype::F16);
+    k.EnableInputDataType(Datatype::F32);
+    k.EnableOutputDataType(Datatype::F16);
+    k.EnableOutputDataType(Datatype::F32);
+    k.EnableInputLayout(DataLayout::bfzyx);
+    k.EnableOutputLayout(DataLayout::bfzyx);
```
Comment (yeonbok) [inline src/plugins/intel_gpu/src/kernel_selector/core/actual_kernels/dft/dft_kernel_ref.cpp:55]: Why does it only support bfzyx?

File: src/plugins/intel_gpu/src/kernel_selector/core/actual_kernels/range/range_kernel_selector.h [hunk 18]
```diff
@@ -4,7 +4,7 @@
 
 #pragma once
 
-#include "kernel_selector.h"
+#include <kernel_selector.h>
```
Comment (yeonbok) [inline src/plugins/intel_gpu/src/kernel_selector/core/actual_kernels/range/range_kernel_selector.h:7]: why range is included from this PR?

File: src/plugins/intel_gpu/src/kernel_selector/core/cl_kernels/dft_ref.cl [hunk 19]
```diff
@@ -0,0 +1,140 @@
...
+#define cadd(a, b) ((cfloat)(real(a) + real(b), imag(a) + imag(b)))
+#define expi(x) ((cfloat)(cos(x), sin(x)))
+#define expmi(x) ((cfloat)(cos(x), -sin(x)))
+#define cload(p, offset, pitch) ((cfloat)((p)[offset], (p)[(offset)+(pitch)]))
+#define cstore(p, offset, pitch, x) ((p)[offset] = real(x), (p)[(offset)+(pitch)] = imag(x))
+#define czero() ((cfloat)(0))
+
+// TODO: pregenerate e{r,i} array on host in macro. maybe it could be done with kernel which runs once?
+KERNEL (dft_ref)(const __global INPUT0_TYPE *data, __global OUTPUT_TYPE *output)
+{
```
Comment (yeonbok) [inline src/plugins/intel_gpu/src/kernel_selector/core/cl_kernels/dft_ref.cl:20]: Please update the cl kernels according to the clang format, i.e., 
"func(~~~) { 
}"

File: src/plugins/intel_gpu/src/kernel_selector/core/common/primitive_db.h [hunk 20]
```diff
@@ -22,7 +22,7 @@ struct primitive_db {
     primitive_db();
 
     std::vector<code> get(const primitive_id& id) const;
-    std::vector<code> get_batch_header_str() const { return std::move(batch_header_str); }
+    std::vector<code> get_batch_header_str() const { return batch_header_str; }
```
Comment (yeonbok) [inline src/plugins/intel_gpu/src/kernel_selector/core/common/primitive_db.h:25]: why did you fixed this?

File: src/plugins/intel_gpu/src/kernel_selector/core/kernel_selector_params.cpp [hunk 21]
```diff
@@ -414,15 +414,13 @@ bool ParamsKey::Support(const ParamsKey& k) const {
         return false;
     if (!((key.outputWeightsType.raw & k.key.outputWeightsType.raw) == k.key.outputWeightsType.raw))
         return false;
-    if (!((key.inputLayout & k.key.inputLayout) != 0 || key.inputLayout == k.key.inputLayout))
+    if (!(k.key.inputLayout.none() || (key.inputLayout & k.key.inputLayout).any()))
```
Comment (yeonbok) [inline src/plugins/intel_gpu/src/kernel_selector/core/kernel_selector_params.cpp:417]: why did you fixed this? please do not change common code if it is not critical for adding a new primitive...

**Comment by wilson-seok** [comment]
@serhii-pavlovskyi-altran Please add clDNN unit test for DFT/IDFT.

File: src/tests/functional/shared_test_classes/src/single_layer/dft.cpp (hunk 13 above)
Comment (mhnap-lohika) [reply src/tests/functional/shared_test_classes/src/single_layer/dft.cpp]: Not really, seems that '.' causes some problems for parsing in Eclipse IDE, but we have the same logic for all other tests, and I don't see a big problem here, reverted.

File: src/tests/functional/shared_test_classes/src/single_layer/dft.cpp (hunk 14 above)
Comment (mhnap-lohika) [reply src/tests/functional/shared_test_classes/src/single_layer/dft.cpp:43]: This is a member of LayerTestsCommon base class. We need this because the default 1e-2f threshold is not enough for these tests.

File: src/plugins/intel_gpu/src/kernel_selector/core/actual_kernels/range/range_kernel_selector.h (hunk 18 above)
Comment (mhnap-lohika) [reply src/plugins/intel_gpu/src/kernel_selector/core/actual_kernels/range/range_kernel_selector.h]: Seems that this is only a small refactoring of another operation and no real logic is changed, I have reverted this.

File: src/plugins/intel_gpu/src/kernel_selector/core/common/primitive_db.h (hunk 20 above)
Comment (mhnap-lohika) [reply src/plugins/intel_gpu/src/kernel_selector/core/common/primitive_db.h]: In this case, std::move is useless, as this method is marked as const. So probably, we should remove const or remove std::move. But still, this is not the scope of this PR, I have reverted it.

File: src/plugins/intel_gpu/include/intel_gpu/primitives/dft.hpp (hunk 16 above)
Comment (mhnap-lohika) [reply src/plugins/intel_gpu/include/intel_gpu/primitives/dft.hpp]: Done

File: src/plugins/intel_gpu/src/plugin/ops/dft.cpp (hunk 11 above)
Comment (mhnap-lohika) [reply src/plugins/intel_gpu/src/plugin/ops/dft.cpp]: done

File: src/plugins/intel_gpu/src/kernel_selector/core/actual_kernels/dft/dft_kernel_ref.cpp (hunk 17 above)
Comment (mhnap-lohika) [reply src/plugins/intel_gpu/src/kernel_selector/core/actual_kernels/dft/dft_kernel_ref.cpp:59]: Added other dimensions.

**Comment by mhnap-lohika** [comment]
@yeonbok @wilson-seok @tgubanova-lohika Seems that I have finished with all comments/propositions and this PR is ready for review.

**Comment by wilson-seok** [comment]
> @yeonbok @wilson-seok @tgubanova-lohika Seems that I have finished with all comments/propositions and this PR is ready for review.

Have you checked single layer test result? I got 16 failed test cases after cherry pick the commits on the latest master. Please check it.
image

**Comment by mhnap-lohika** [comment]
> > @yeonbok @wilson-seok @tgubanova-lohika Seems that I have finished with all comments/propositions and this PR is ready for review.
> 
> Have you checked single layer test result? I got 16 failed test cases after cherry pick the commits on the latest master. Please check it. CI check ie_test_gpu_ubuntu18_release also has same DFT test failure cased. image

Strange, all tests are good for me, but I have Ubuntu 20. Will try to investigate why 18 has problems.
image

**Comment by wilson-seok** [comment]
> > > @yeonbok @wilson-seok @tgubanova-lohika Seems that I have finished with all comments/propositions and this PR is ready for review.
> > 
> > 
> > Have you checked single layer test result? I got 16 failed test cases after cherry pick the commits on the latest master. Please check it. CI check ie_test_gpu_ubuntu18_release also has same DFT test failure cased. image
> 
> Strange, all tests are good for me, but I have Ubuntu 20. I don't have access to ie_test_gpu_ubuntu18_release in CI, but still will try to investigate why there are problems with Ubuntu 18. image

Could you try the test after rebasing to the latest master? In my local, I cherry-picked this PR on the latest master and reproduced the issue in Ubuntu20.

**Comment by wilson-seok** [comment]
> > > > > @yeonbok @wilson-seok @tgubanova-lohika Seems that I have finished with all comments/propositions and this PR is ready for review.
> > > > 
> > > > 
> > > > Have you checked single layer test result? I got 16 failed test cases after cherry pick the commits on the latest master. Please check it. CI check ie_test_gpu_ubuntu18_release also has same DFT test failure cased. image
> > > 
> > > 
> > > Strange, all tests are good for me, but I have Ubuntu 20. I don't have access to ie_test_gpu_ubuntu18_release in CI, but still will try to investigate why there are problems with Ubuntu 18. image
> > 
> > 
> > Could you try the test after rebasing to the latest master? In my local, I cherry-picked this PR on the latest master and reproduced the issue in Ubuntu20.
> 
> Yeah, I have rebased to the latest master, and all GPU (I)DFT tests are passing for me. Maybe I need to turn on/off some specific flags in build options to be able to reproduce it?

I checked cherry-pick with yesterday rebase commits and now the issue disappeared. ie_test_gpu_ubuntu18_release is passed also.

File: src/plugins/intel_gpu/src/kernel_selector/core/actual_kernels/dft/dft_kernel_ref.cpp (hunk 17 above)
Comment (wilson-seok) [reply src/plugins/intel_gpu/src/kernel_selector/core/actual_kernels/dft/dft_kernel_ref.cpp:117]: Could you please add those layout for input/output? Those are required for dGPU support.
<html>
<body>


Prec | Layout
-- | --
int8 | fsv16, fsv32, bsv32_fsv32
fp16 | fsv16, bsv32_fsv16
fp32 | fsv16, bsv32_fsv16


</body>
</html>

**Comment by mhnap-lohika** [comment]
> > > > > > @yeonbok @wilson-seok @tgubanova-lohika Seems that I have finished with all comments/propositions and this PR is ready for review.
> > > > > 
> > > > > 
> > > > > Have you checked single layer test result? I got 16 failed test cases after cherry pick the commits on the latest master. Please check it. CI check ie_test_gpu_ubuntu18_release also has same DFT test failure cased. image
> > > > 
> > > > 
> > > > Strange, all tests are good for me, but I have Ubuntu 20. I don't have access to ie_test_gpu_ubuntu18_release in CI, but still will try to investigate why there are problems with Ubuntu 18. image
> > > 
> > > 
> > > Could you try the test after rebasing to the latest master? In my local, I cherry-picked this PR on the latest master and reproduced the issue in Ubuntu20.
> > 
> > 
> > Yeah, I have rebased to the latest master, and all GPU (I)DFT tests are passing for me. Maybe I need to turn on/off some specific flags in build options to be able to reproduce it?
> 
> I checked cherry-pick with yesterday rebase commits and now the issue disappeared. ie_test_gpu_ubuntu18_release is passed also.

Great news! So I think we can proceed with the review.

File: src/tests/functional/shared_test_classes/src/single_layer/dft.cpp (hunk 14 above)
Comment (yeonbok) [reply src/tests/functional/shared_test_classes/src/single_layer/dft.cpp:43]: Could you explain why the accuracy could not meet?

File: src/plugins/intel_gpu/include/intel_gpu/primitives/dft.hpp (hunk 16 above)
Comment (yeonbok) [reply src/plugins/intel_gpu/include/intel_gpu/primitives/dft.hpp]: Its still here and seems no need?

**Comment by wilson-seok** [comment]
Please rebase the PR to resolve issue in openvino-mac check.

File: src/plugins/intel_gpu/tests/test_cases/dft_gpu_test.cpp [hunk 22]
```diff
@@ -0,0 +1,1591 @@
...
+
+TEST_P(idft_gpu_test_f32, idft_f32) {
+    test();
+}
+
+TEST_P(idft_gpu_test_f16, idft_f16) {
+    test();
+}
+
+const std::vector<float> input_data = {
```
Comment (lohika-denis-kotov) [inline src/plugins/intel_gpu/tests/test_cases/dft_gpu_test.cpp:155]: Looks like this blob vectors would be better to store in separate file and include here using `#include`, because it make this file bigger and harder to read the logic in `dft_gpu_test`
If you want to see in this scope `const std::vector<float> input_data` then it is better as for me to declare it like this `extern const std::vector<float> input_data`:
```cpp
extern const std::vector<float> input_data;
#include "dft_gpu_test_blobls.hpp"
```

File: src/tests/functional/shared_test_classes/src/single_layer/dft.cpp [hunk 23]
```diff
@@ -26,22 +26,20 @@ std::string DFTLayerTest::getTestCaseName(const testing::TestParamInfo<DFTParams
...
 
 void DFTLayerTest::SetUp() {
-    InferenceEngine::SizeVector inputShapes;
+    InferenceEngine::SizeVector inputShape;
     InferenceEngine::Precision inputPrecision;
     std::vector<int64_t> axes;
     std::vector<int64_t> signalSize;
     ngraph::helpers::DFTOpType opType;
-    std::tie(inputShapes, inputPrecision, axes, signalSize, opType, targetDevice) = this->GetParam();
+    tie(inputShape, inputPrecision, axes, signalSize, opType, targetDevice) = GetParam();
```
Comment (lohika-denis-kotov) [inline src/tests/functional/shared_test_classes/src/single_layer/dft.cpp:34]: Now this code is reliy on ADL ...
What is `InferenceEngine::SizeVector` type will be changed ? It will potentially break search of tie function in `std` namespace ...

I personally prefer to use `std::tie`

File: src/tests/functional/shared_test_classes/src/single_layer/dft.cpp (hunk 14 above)
Comment (mhnap-lohika) [reply src/tests/functional/shared_test_classes/src/single_layer/dft.cpp:43]: This can be because ngraph reference implementation use FFT as (I)DFT implementation and these methods can have different precisions.

File: src/plugins/intel_gpu/tests/test_cases/dft_gpu_test.cpp (hunk 22 above)
Comment (mhnap-lohika) [reply src/plugins/intel_gpu/tests/test_cases/dft_gpu_test.cpp:155]: Yeah, nice idea. I took those tests data from ngraph reference tests, and also was thinking about extracting them somewhere. But agreed with @tgubanova-lohika that we won't do it for now as it will cause dependencies if we extract them together with ngraph and won't make sense to extract only in cldnn.

File: src/tests/functional/shared_test_classes/src/single_layer/dft.cpp (hunk 23 above)
Comment (mhnap-lohika) [reply src/tests/functional/shared_test_classes/src/single_layer/dft.cpp]: reverted those changes, I also don't like using without std

File: src/plugins/intel_gpu/src/kernel_selector/core/actual_kernels/dft/dft_kernel_ref.cpp (hunk 17 above)
Comment (mhnap-lohika) [reply src/plugins/intel_gpu/src/kernel_selector/core/actual_kernels/dft/dft_kernel_ref.cpp:117]: agreed not to do for now

**Comment by mhnap-lohika** [comment]
@yeonbok I made smaller dimensions in single layer tests, so now we don't need to change the default threshold, so I reverted those Serhii changes. Would be great if @wilson-seok can run tests once more.

**Comment by wilson-seok** [comment]
> @yeonbok I made smaller dimensions in single layer tests, so now we don't need to change the default threshold, so I reverted those Serhii changes. Would be great if @wilson-seok can run tests once more.

On my local ubuntu20 machine, clDNN unit test and single layer test passed all. But in ie_tests_win_gpu_vs2019_release, there are 2 failed cases in single layer test. Please check them in Windows.

image
image

**Comment by mhnap-lohika** [comment]
> > @yeonbok I made smaller dimensions in single layer tests, so now we don't need to change the default threshold, so I reverted those Serhii changes. Would be great if @wilson-seok can run tests once more.
> 
> On my local ubuntu20 machine, clDNN unit test and single layer test passed all. But in ie_tests_win_gpu_vs2019_release, there are 2 failed cases in single layer test. Please check them in Windows.
> 
> image image

Tweaked tests a little bit more, should pass now.

**Review by wilson-seok — APPROVED** [review]
LGTM

File: src/plugins/intel_gpu/src/kernel_selector/core/cl_kernels/dft_ref.cl [hunk 24]
```diff
@@ -0,0 +1,139 @@
+// Copyright (C) 2022 Intel Corporation
+// SPDX-License-Identifier: Apache-2.0
+//
+
+#define GET_INDEX(prefix, ORDER) CAT(prefix, _GET_INDEX)(ORDER)
+
+// alternative: https://github.com/OpenCL/ComplexMath/blob/master/clcomplex.h
+typedef float2 cfloat;
```
Comment (yeonbok) [inline src/plugins/intel_gpu/src/kernel_selector/core/cl_kernels/dft_ref.cl:8]: Please do not redefine this here, Because the cl codes are combined altogether and redefining here might cause error in the succeeding kernels expectation.

**Review by yeonbok — APPROVED** [review]
